pkg/pipeline/DeploymentConfigService.go

```diff
@@ -0,0 +1,267 @@
+			DisplayName: "Strategy configuration",
+			Value:       pipelineStrategy.Config,
+		},
+	}
+	return pipelineStrategyConfig, nil
+}
+
+func (impl *DeploymentConfigServiceImpl) GetLatestCMCSConfig(pipeline *pipelineConfig.Pipeline, userHasAdminAccess bool) ([]*history.ComponentLevelHistoryDetailDto, []*history.ComponentLevelHistoryDetailDto, error) {
+
+	configAppLevel, err := impl.configMapRepository.GetByAppIdAppLevel(pipeline.AppId)
+	if err != nil && pg.ErrNoRows != err {
+		impl.logger.Errorw("error in getting CM/CS app level data", "err", err, "appId", pipeline.AppId)
+		return nil, nil, err
+	}
+	var configMapAppLevel string
+	var secretAppLevel string
+	if configAppLevel != nil && configAppLevel.Id > 0 {
+		configMapAppLevel = configAppLevel.ConfigMapData
+		secretAppLevel = configAppLevel.SecretData
+	}
+	configEnvLevel, err := impl.configMapRepository.GetByAppIdAndEnvIdEnvLevel(pipeline.AppId, pipeline.EnvironmentId)
+	if err != nil && pg.ErrNoRows != err {
+		impl.logger.Errorw("error in getting CM/CS env level data", "err", err, "appId", pipeline.AppId)
+		return nil, nil, err
+	}
+	var configMapEnvLevel string
+	var secretEnvLevel string
+	if configEnvLevel != nil && configEnvLevel.Id > 0 {
+		configMapEnvLevel = configEnvLevel.ConfigMapData
+		secretEnvLevel = configEnvLevel.SecretData
+	}
+	mergedConfigMap, err := impl.MergeCMCSConfig(configMapAppLevel, configMapEnvLevel, repository2.CONFIGMAP_TYPE)
+	if err != nil {
+		impl.logger.Errorw("error in merging app level and env level CM configs", "err", err)
+		return nil, nil, err
+	}
+	mergedSecret, err := impl.MergeCMCSConfig(secretAppLevel, secretEnvLevel, repository2.SECRET_TYPE)
+	if err != nil {
+		impl.logger.Errorw("error in merging app level and env level CM configs", "err", err)
+		return nil, nil, err
+	}
+
+	var cmConfigsDto []*history.ComponentLevelHistoryDetailDto
+	for _, data := range mergedConfigMap {
+		convertedData, err := impl.configMapHistoryService.ConvertConfigDataToComponentLevelDto(data, repository2.CONFIGMAP_TYPE, userHasAdminAccess)
+		if err != nil {
+			impl.logger.Errorw("error in converting cmConfig to componentLevelData", "err", err)
+			return nil, nil, err
+		}
+		cmConfigsDto = append(cmConfigsDto, convertedData)
+	}
+
+	var secretConfigsDto []*history.ComponentLevelHistoryDetailDto
+	for _, data := range mergedSecret {
+		convertedData, err := impl.configMapHistoryService.ConvertConfigDataToComponentLevelDto(data, repository2.SECRET_TYPE, userHasAdminAccess)
+		if err != nil {
+			impl.logger.Errorw("error in converting secretConfig to componentLevelData", "err", err)
+			return nil, nil, err
+		}
+		secretConfigsDto = append(secretConfigsDto, convertedData)
+	}
+	return cmConfigsDto, secretConfigsDto, nil
+}
+
+func (impl *DeploymentConfigServiceImpl) MergeCMCSConfig(appLevelConfig, envLevelConfig string, configType repository2.ConfigType) (map[string]*history.ConfigData, error) {
+	envLevelMap := make(map[string]*history.ConfigData, 0)
```

Comment:
are we duplicating this logic ?

Reply:
no, different objects are used in this and